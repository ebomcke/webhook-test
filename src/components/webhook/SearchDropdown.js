import React, { Component } from 'react';
import { compose } from 'recompose';
import { withEndpoints } from '../../contexts/Endpoints';
import { Dropdown, Label, LabelDetail, Icon, Segment } from 'semantic-ui-react';

class SearchDropdown extends Component {
  constructor(props) {
    super(props);
    let options = props.endpoints.map(endpoint => ({
      key: endpoint.id,
      text: `url: /${endpoint.path}`,
      value: endpoint.id,
      type: 'path',
      filtervalue: endpoint.path,
    }));
    const { filters } = this.props;
    const contentFilters = filters.filter(f => f.type === 'content');
    if (contentFilters.length > 0) {
      options = options.concat(contentFilters.map(f => this.generateContentOption(f.value)));
    }
    this.state = {
      options,
    };
  }

  generateContentOption = (value) => ({
    key: `content-${value}`,
    text: `content: ${value}`,
    value,
    type: 'content',
    filtervalue: value,
  })

  addContentSearch = (e, { value }) => {
    const { options } = this.state;
    if (!options.find(o => o.value === value)) {
      options.push(this.generateContentOption(value));
      this.setState({ options });
      const { filters, updateFilters } = this.props;
      updateFilters(
        filters.concat([
          {
            type: 'content',
            value: value,
          },
        ]),
      );
    }
  };

  handleChange = (e, { value: values }) => {
    const { options } = this.state;
    const filters = values
      .map(value => options.find(o => o.value === value))
      .map(
        o =>
          o && {
            type: o.type,
            value: o.filtervalue,
          },
      )
      .filter(f => f);
    this.props.updateFilters(filters);
  };

  render() {
    const { options } = this.state;
    const { filters } = this.props;
    const defaultValue = filters.map(
      f =>
        options.find(o => o.type === f.type && o.filtervalue === f.value).value,
    );
    return (
      <Dropdown
        fluid
        multiple
        search
        selection
        placeholder="Filter by content and/or url"
        allowAdditions={true}
        defaultValue={defaultValue}
        additionLabel="content: "
        onAddItem={this.addContentSearch}
        closeOnChange={true}
        openOnFocus={false}
        icon="search"
        options={options}
        onChange={this.handleChange}
      />
    );
  }
}

export default compose(withEndpoints)(SearchDropdown);
