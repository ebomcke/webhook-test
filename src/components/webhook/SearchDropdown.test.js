import React from 'react';
import { create } from 'react-test-renderer';
import { Dropdown } from 'semantic-ui-react';
import SearchDropdown from './SearchDropdown';

const filterContent = {
  type: 'content',
  value: 'unit-test',
};
const filterEndpoint = {
  type: 'path',
  value: 'unit-test-1',
};

describe('Rendering tests', () => {
  test('No filters', () => {
    const searchDropdown = create(<SearchDropdown filters={[]} />);
    expect(searchDropdown.toJSON()).toMatchSnapshot();
  });

  test('Content filter', () => {
    const searchDropdown = create(<SearchDropdown filters={[filterContent]} />);
    expect(searchDropdown.toJSON()).toMatchSnapshot();
  });

  test('Path filter', () => {
    const searchDropdown = create(
      <SearchDropdown filters={[filterEndpoint]} />,
    );
    expect(searchDropdown.toJSON()).toMatchSnapshot();
  });

  test('Both filters', () => {
    const searchDropdown = create(
      <SearchDropdown filters={[filterEndpoint, filterContent]} />,
    );
    expect(searchDropdown.toJSON()).toMatchSnapshot();
  });
});

describe('Interactions tests', () => {
  let searchDropdown;
  const updateFilters = jest.fn();

  beforeEach(() => {
    searchDropdown = create(
      <SearchDropdown
        filters={[filterEndpoint, filterContent]}
        updateFilters={updateFilters}
      />,
    );
    updateFilters.mockClear();
  });

  test('Select endpoint adds an endpoint filter and preserves content filters only', () => {
    searchDropdown.root.findByType(Dropdown).props.onChange(null, {
      value: ['endpoint-2', 'unit-test'],
    });
    expect(updateFilters.mock.calls.length).toBe(1);
    expect(updateFilters.mock.calls[0][0]).toMatchSnapshot();
  });

  test('Add item adds a content filter and preserves existing filters', () => {
    searchDropdown.root.findByType(Dropdown).props.onAddItem(null, {
      value: 'new-filter',
    });
    expect(updateFilters.mock.calls.length).toBe(1);
    expect(updateFilters.mock.calls[0][0]).toMatchSnapshot();
  });

  test('Remove endpoint updates filters', () => {
    // Mock event and value as semantic UI library relies on it
    const label = searchDropdown.root.findByProps({ value: 'endpoint-1' });
    label.findByProps({ className: 'delete icon' }).props.onClick(
      {
        stopPropagation: jest.fn(),
      },
      {
        value: 'endpoint-1',
      },
    );
    expect(updateFilters.mock.calls.length).toBe(1);
    expect(updateFilters.mock.calls[0][0]).toMatchSnapshot();
  });

  test('Remove content updates filters', () => {
    // Mock event and value as semantic UI library relies on it
    const label = searchDropdown.root.findByProps({ value: 'unit-test' });
    label.findByProps({ className: 'delete icon' }).props.onClick(
      {
        stopPropagation: jest.fn(),
      },
      {
        value: 'endpoint-1',
      },
    );
    expect(updateFilters.mock.calls.length).toBe(1);
    expect(updateFilters.mock.calls[0][0]).toMatchSnapshot();
  });
});
