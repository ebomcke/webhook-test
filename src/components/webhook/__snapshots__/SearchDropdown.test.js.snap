// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interactions tests Add item adds a content filter and preserves existing filters 1`] = `
Array [
  Object {
    "type": "path",
    "value": "unit-test-1",
  },
  Object {
    "type": "content",
    "value": "unit-test",
  },
  Object {
    "type": "content",
    "value": "new-filter",
  },
]
`;

exports[`Interactions tests Remove content updates filters 1`] = `
Array [
  Object {
    "type": "path",
    "value": "unit-test-1",
  },
]
`;

exports[`Interactions tests Remove endpoint updates filters 1`] = `
Array [
  Object {
    "type": "content",
    "value": "unit-test",
  },
]
`;

exports[`Interactions tests Select endpoint adds an endpoint filter and preserves content filters only 1`] = `
Array [
  Object {
    "type": "path",
    "value": "unit-test-2",
  },
  Object {
    "type": "content",
    "value": "unit-test",
  },
]
`;

exports[`Rendering tests Both filters 1`] = `
<div
  aria-expanded={false}
  className="ui fluid multiple search selection dropdown"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  role="combobox"
>
  <a
    className="ui label"
    onClick={[Function]}
    value="endpoint-1"
  >
    url: /unit-test-1
    <i
      aria-hidden="true"
      className="delete icon"
      onClick={[Function]}
    />
  </a>
  <a
    className="ui label"
    onClick={[Function]}
    value="unit-test"
  >
    content: unit-test
    <i
      aria-hidden="true"
      className="delete icon"
      onClick={[Function]}
    />
  </a>
  <input
    aria-autocomplete="list"
    autoComplete="off"
    className="search"
    onChange={[Function]}
    style={
      Object {
        "width": undefined,
      }
    }
    tabIndex={0}
    type="text"
    value=""
  />
  <span
    className="sizer"
  />
  <div
    aria-live="polite"
    className="text"
    role="alert"
  />
  <i
    aria-hidden="true"
    className="search icon"
    onClick={[Function]}
  />
  <div
    aria-multiselectable={true}
    className="menu transition"
    role="listbox"
  >
    <div
      aria-checked={false}
      aria-selected={true}
      className="selected item"
      filtervalue="unit-test-2"
      onClick={[Function]}
      role="option"
      style={
        Object {
          "pointerEvents": "all",
        }
      }
      type="path"
    >
      <span
        className="text"
      >
        url: /unit-test-2
      </span>
    </div>
  </div>
</div>
`;

exports[`Rendering tests Content filter 1`] = `
<div
  aria-expanded={false}
  className="ui fluid multiple search selection dropdown"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  role="combobox"
>
  <a
    className="ui label"
    onClick={[Function]}
    value="unit-test"
  >
    content: unit-test
    <i
      aria-hidden="true"
      className="delete icon"
      onClick={[Function]}
    />
  </a>
  <input
    aria-autocomplete="list"
    autoComplete="off"
    className="search"
    onChange={[Function]}
    style={
      Object {
        "width": undefined,
      }
    }
    tabIndex={0}
    type="text"
    value=""
  />
  <span
    className="sizer"
  />
  <div
    aria-live="polite"
    className="text"
    role="alert"
  />
  <i
    aria-hidden="true"
    className="search icon"
    onClick={[Function]}
  />
  <div
    aria-multiselectable={true}
    className="menu transition"
    role="listbox"
  >
    <div
      aria-checked={false}
      aria-selected={true}
      className="selected item"
      filtervalue="unit-test-1"
      onClick={[Function]}
      role="option"
      style={
        Object {
          "pointerEvents": "all",
        }
      }
      type="path"
    >
      <span
        className="text"
      >
        url: /unit-test-1
      </span>
    </div>
    <div
      aria-checked={false}
      aria-selected={false}
      className="item"
      filtervalue="unit-test-2"
      onClick={[Function]}
      role="option"
      style={
        Object {
          "pointerEvents": "all",
        }
      }
      type="path"
    >
      <span
        className="text"
      >
        url: /unit-test-2
      </span>
    </div>
  </div>
</div>
`;

exports[`Rendering tests No filters 1`] = `
<div
  aria-expanded={false}
  className="ui fluid multiple search selection dropdown"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  role="combobox"
>
  <input
    aria-autocomplete="list"
    autoComplete="off"
    className="search"
    onChange={[Function]}
    style={
      Object {
        "width": undefined,
      }
    }
    tabIndex={0}
    type="text"
    value=""
  />
  <span
    className="sizer"
  />
  <div
    aria-live="polite"
    className="default text"
    role="alert"
  >
    Filter by content and/or url
  </div>
  <i
    aria-hidden="true"
    className="search icon"
    onClick={[Function]}
  />
  <div
    aria-multiselectable={true}
    className="menu transition"
    role="listbox"
  >
    <div
      aria-checked={false}
      aria-selected={true}
      className="selected item"
      filtervalue="unit-test-1"
      onClick={[Function]}
      role="option"
      style={
        Object {
          "pointerEvents": "all",
        }
      }
      type="path"
    >
      <span
        className="text"
      >
        url: /unit-test-1
      </span>
    </div>
    <div
      aria-checked={false}
      aria-selected={false}
      className="item"
      filtervalue="unit-test-2"
      onClick={[Function]}
      role="option"
      style={
        Object {
          "pointerEvents": "all",
        }
      }
      type="path"
    >
      <span
        className="text"
      >
        url: /unit-test-2
      </span>
    </div>
  </div>
</div>
`;

exports[`Rendering tests Path filter 1`] = `
<div
  aria-expanded={false}
  className="ui fluid multiple search selection dropdown"
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  role="combobox"
>
  <a
    className="ui label"
    onClick={[Function]}
    value="endpoint-1"
  >
    url: /unit-test-1
    <i
      aria-hidden="true"
      className="delete icon"
      onClick={[Function]}
    />
  </a>
  <input
    aria-autocomplete="list"
    autoComplete="off"
    className="search"
    onChange={[Function]}
    style={
      Object {
        "width": undefined,
      }
    }
    tabIndex={0}
    type="text"
    value=""
  />
  <span
    className="sizer"
  />
  <div
    aria-live="polite"
    className="text"
    role="alert"
  />
  <i
    aria-hidden="true"
    className="search icon"
    onClick={[Function]}
  />
  <div
    aria-multiselectable={true}
    className="menu transition"
    role="listbox"
  >
    <div
      aria-checked={false}
      aria-selected={true}
      className="selected item"
      filtervalue="unit-test-2"
      onClick={[Function]}
      role="option"
      style={
        Object {
          "pointerEvents": "all",
        }
      }
      type="path"
    >
      <span
        className="text"
      >
        url: /unit-test-2
      </span>
    </div>
  </div>
</div>
`;
